### PR TITLE
Add Mailspring Flatpak

### PR DESCRIPTION
**Pull request by kirbyfan64**
[Related discussion.](https://github.com/Foundry376/Mailspring/issues/345) (FWIW the sync engine license explicitly allows redistribution.)

### CODE REVIEW

**Comment by nedrichards** [comment]
What's the justification for not compiling from source? I can see possible issues with the closed source component and npm, but I'd like to see them confirmed.

**Comment by nedrichards** [comment]
Could you send the appdata upstream? This sort of thing isn't meant to live only in flathub, but be shared by all.

**Comment by nedrichards** [comment]
when compiling locally:
> sh: 1: sh: rpm2cpio: not found
cpio: premature end of archive
Error: module mailspring: Child process exited with code 1

**Comment by kirbyfan64** [comment]
@nedrichards Well I was going to submit the appstream metadata, but I realized the XDG desktop file isn't in the repository either (only the binaries), so I filed [this issue](https://github.com/Foundry376/Mailspring/issues/1069).

The rpm2cpio error is because I used the .rpm file as a source, and flatpak-builder needs rpm2cpio in order to extract .rpm files.

**Comment by nedrichards** [comment]
Good to get that cleared up up upstream.

What's your plan with rpm2cpio then?

Is there a plan for compiling from source (even including the closed source part as a binary module) that really is *significantly* more user friendly that extra-data.

**Comment by kirbyfan64** [comment]
@nedrichards It was admittedly partly laziness, since no other Flatpak builds Electron apps from source (not even Atom). IMO rpm2cpio not being a required dependency is kind of flatpak-builder's problem, since I'm using a normal, documented feature in the intended way... :sweat_smile: (Also since distribution is allowed according to Mailspring's license, I'm not using any extra-data.)

If rpm2cpio concerns you, I can look into using the .deb file instead.

**Comment by nedrichards** [comment]
@kirbyfan64 Your build occurs inside the GNOME SDK which doesn't contain rpm2cpio, if you want it to then you have to compile it in the recipe then clean it up yourself, or extract with the tools you get in the SDK another way.

There are apps that build electron from source. Here's one of them: https://github.com/flathub/com.visualstudio.code.oss

**Comment by sramkrishna** [comment]
So what's the status of this? :-)

**Comment by barthalion** [comment]
As label says, awaiting changes. Also unclear upstream involvement.

**Comment by kirbyfan64** [comment]
Sorry, no progress yet. Tried to change it to use the .deb package and got momentarily stuck.

**Comment by kirbyfan64** [comment]
@nedrichards A pull request has been opened to add the appdata upstream. Meanwhile, I've made it work from the Ubuntu deb package instead (so there shouldn't be any rpm2cpio issues), as well as adding PulseAudio support like @cungureanu said.

File: com.getmailspring.Mailspring.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ app-id: com.getmailspring.Mailspring
 base: org.electronjs.Electron2.BaseApp
 base-version: stable
 runtime: org.gnome.Platform
-runtime-version: 3.28
+runtime-version: 3.30
```
Comment (bilelmoussaoui) [inline com.getmailspring.Mailspring.yaml]: This is a string. Needs a quote or a double quote

File: com.getmailspring.Mailspring.yaml [hunk 2]
```diff
@@ -2,7 +2,7 @@ app-id: com.getmailspring.Mailspring
 base: org.electronjs.Electron2.BaseApp
 base-version: stable
 runtime: org.gnome.Platform
```
Comment (bilelmoussaoui) [inline com.getmailspring.Mailspring.yaml]: Is there a specific reason for using the GNOME platform?

**Comment by kirbyfan64** [comment]
@bilelmoussaoui Done.

> Is there a specific reason for using the GNOME platform?

For libsecret and libappindicator.

**Comment by nedrichards** [comment]
bot, build com.getmailspring.Mailspring

**Comment by cungureanu** [comment]
Why it still has awaiting-changes label?

**Comment by aiamuzz** [comment]
i am waiting for it too ... very eagerly ...

**Comment by haevalencia** [comment]
👀

**Comment by nedrichards** [comment]
From my perspective - this is awaiting an answer to the 'what reason can you not build from source' question above https://github.com/flathub/flathub/pull/592#issuecomment-420016073

I assume from the somewhat acrimounious discussion linked at the top that upstream isn't interested in maintaining this themselves.

**Comment by kirbyfan64** [comment]
I can try and do a source build, but it'll take a bit for me to have time...

Meanwhile AFAIK every other complaint/issue has been addressed, and I'll again point out there are other Electron-based Flatpaks for OSS repos that still build from binaries...

**Comment by nedrichards** [comment]
bot, build com.getmailspring.Mailspring

And yes, there are other electron based flatpaks of oss software that build from binaries. In all of the cases that have happened recently that's because the existing tooling isn't sufficient to build them (e.g. one of their dependencies absolutely requires downloading a binary component during the build, which is blocked by the build sandbox). That's what I mean by 'what reason aren't you building from source'. If you have a compelling reason then it can be considered but it's not a go to.

**Comment by nedrichards** [comment]
And note, there's an appdata wiki page now that should help with common problems: https://github.com/flathub/flathub/wiki/AppData-Guidelines

**Comment by nedrichards** [comment]
Yep, built this locally and looked at the compiled appdata in /app/share/app-info/xmls/com.getmailspring.Mailspring.xml.gz when you extract it and view you can see that it says:

> <id>mailspring</id>

where it should say:

> <id>com.getmailspring.Mailspring</id>

With that patch to the appdata, it should be fine.

**Comment by kirbyfan64** [comment]
I have no time/desire for this right now and probably won't get enough time to do figure out a source build for a while.

**Comment by aiamuzz** [comment]
hmmmmm ... was looking forward to this app getting added to flathub ... :O

**Comment by kirbyfan64** [comment]
FWIW I might pick this up again later on, I just don't have time right now.

**Comment by aiamuzz** [comment]
> FWIW I might pick this up again later on, I just don't have time right now.

Hey ... !!!

**Pull request closed without merging**